Objects/TI4_Helpers/TI4_FactionHelper.ttslua

```diff
@@ -1122,6 +1126,703 @@ function tintTokens(params)
+            local qy = u1m1sqrt *math.cos(2*math.pi*u2);
+            local qz = u1sqrt *math.sin(2*math.pi*u3);
+            local qw = u1sqrt *math.cos(2*math.pi*u3);
+            --Apply rotation
+            local ysqr = qy * qy;
+            local t0 = -2.0 * (ysqr + qz * qz) + 1.0;
+            local t1 = 2.0 * (qx * qy - qw * qz);
+            local t2 = -2.0 * (qx * qz + qw * qy);
+            local t3 = 2.0 * (qy * qz - qw * qx);
+            local t4 = -2.0 * (qx * qx + ysqr) + 1.0;
+            --Correct
+            if t2 > 1.0 then t2 = 1.0 end
+            if t2 < -1.0 then ts = -1.0 end
+            --Convert back to X/Y/Z
+            local xr = math.asin(t2);
+            local yr = math.atan2(t3, t4);
+            local zr = math.atan2(t1, t0);
+            --Return result
+            return {math.deg(xr),math.deg(yr),math.deg(zr)}
+        end
+
+        -- Use different color dice for manual auditability; copy from Multiroller
+        local unitTypeToDiceColor = {
+            ["Dreadnought"] = "Purple",
+            ["Flagship"] = "Black",
+            ["Destroyer"] = "Red",
+            ["War Sun"] = "Orange",
+            ["Carrier"] = "Blue",
+            ["Fighter"] = "Teal",
+            ["Infantry"] = "Green",
+            ["Cruiser"] = "Brown",
+            ["PDS"] = "Orange",
+            ["Space Dock"] = "Yellow",
+            ["Mech"] = "Pink"
+        }
+        
+        local diceToShip = {}
+        local unitTypesUsed = {}
+        _diceToBeRolled = {}
+
+        -- Spawn a die for each ship, around the system tile the ship is on.
+        for system, ships in pairs(systemToAnchoredShips) do
+            local shipCount = #ships
+            local angleStep = (360 / shipCount)
+
+            local systemObject = getObjectFromGUID(system.guid)
+
+            for i, ship in ipairs(ships) do
+                local die = spawnObject({
+                    type=dieType,
+                    position = _findGlobalPosWithLocalDirection(systemObject, angleStep*(i-1)),
+                    rotation = _randomRotation(), scale={dieSize,dieSize,dieSize}
+                })
+                die.setLock(true)
+                die.setColorTint(stringColorToRGB(unitTypeToDiceColor[ship.unitType]))
+
+                diceToShip[die] = ship
+                table.insert(_diceToBeRolled, die)
+                unitTypesUsed[ship.unitType] = true
+            end
+        end
+
+        -- Print guide showing the dice color for each ship.
+        print('DICE COLOR GUIDE')
+        for unitType, color in pairs(unitTypeToDiceColor) do
+            if unitTypesUsed[unitType] then
+                print('* ' .. unitType .. ': ' .. color)
+            end
+        end
+
+        -- Remove dice from board no matter what after 2 minutes
+        local function destroyDice()
+            for _, die in ipairs(_diceToBeRolled) do
+                if getObjectFromGUID(die.getGUID()) then
```

Comment:
The die might not exist, safer to keep around the guid and re-get the object.

Reply:
Will be making a big pass to account for this throughout my changes